notmuch/mutt_notmuch.c

```diff
@@ -778,8 +778,10 @@ static int init_email(struct Email *e, const char *path, notmuch_message_t *msg)
 
   mutt_debug(LL_DEBUG2, "nm: [e=%p, edata=%p] (%s)\n", (void *) e, (void *) e->edata, id);
 
-  if (!e->env->message_id)
-    e->env->message_id = nm2mutt_message_id(id);
+  char *nm_msg_id = nm2mutt_message_id(id);
+  if (!e->env->message_id || (mutt_str_strcmp(e->env->message_id, nm_msg_id) != 0))
+    e->env->message_id = nm_msg_id;
+
```

Comment:
We need to free the `message_id` in the case where it exists, but differs from Notmuch's.

```c
  char *nm_msg_id = nm2mutt_message_id(id);
  if (mutt_str_strcmp(e->env->message_id, nm_msg_id) != 0)
  { 
    FREE(&e->env->message_id);
    e->env->message_id = nm_msg_id;
  }
```

but now I'm worried about `Mailbox->id_hash` (the Mailbox's hash of Message-Ids).
If we pass the `Mailbox` into `init_email()`, we can:

```c
  mutt_hash_delete(m->id_hash, old_message_id, e);  
  mutt_hash_insert(m->id_hash, new_message_id, e);  
```

giving us:

```c
  char *nm_msg_id = nm2mutt_message_id(id);
  if (!e->env->message_id)
  { 
    e->env->message_id = nm_msg_id;
    mutt_hash_insert(m->id_hash, e->env->message_id, e);
  }  
  else if (mutt_str_strcmp(e->env->message_id, nm_msg_id) != 0)
  {
    mutt_hash_delete(m->id_hash, e->env->message_id, e);
    FREE(&e->env->message_id);
    e->env->message_id = nm_msg_id;
    mutt_hash_insert(m->id_hash, e->env->message_id, e);
  }
```

Reply:
I'm not sure the insertions/deletions are necessary.

The hash table is first initialized in `get_mutt_email()`, which is called from `nm_mbox_check()`. `nm_mbox_check()` occurs after the `init_email()` calls. At that point, the hash table is accurate.

If we "delete" an  email by untagging it through `notmuch`'s cli, we receive the "Mailbox was externally modified. Flags may be wrong." message, which is reasonable because it was externally modified. The user should reload their mailbox for accurate information.

If we "add" an email by tagging it through `notmuch`'s cli, we don't receive any message. New messages just appear, which is incorrect behavior. We should warn the user that they should reload.

If we do proceed with inserting/deleting, we will need to initialize the hash table in `init_email()`.

Reply:
OK, fair enough.
Then we just need the `FREE()` for the replacement case.

Reply:
I've added a `FREE()` for the replacement case and another for when we don't use `nm_msg_id` since `nm2mutt_message_id()` has a malloc call.